### PR TITLE
Overhaul testing

### LINKED ISSUE #13: Update to latest release of netcdf-c (v4.9.0)

### LINKED ISSUE #30: Use nextSIM-DG Docker build for CI
We have a Docker build with the dependencies of nextSIM-DG installed. If we add catch2, trilinos, and zoltan to this then we'll be able to use it here, too. It'd cut down runtime for the CI test suite significantly.

### LINKED ISSUE #65: Move from Catch2 to doctest
Catch2 is proving to be a difficult dependency to handle. Given that doctest is "heavily influenced by Catch" and it's used for the testing framework in nextSIM-DG, it would make sense to use it here, too.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #13.
Closes #30.
Closes #65.

This PR started out as a simple version update for `netcdf-c` but ended up being a complete overhaul of the testing. Catch2 is a difficult dependency to handle, so I propose switching to doctest, which is already in use in nextSIM-DG. By doing so, we are able to make use of the nextSIM-DG Docker build for the Ubuntu CI job.

A limitation of doctest is that (it seems to me) you can't have tests with different numbers of MPI ranks in the same file, so I've separated them out. It introduces some duplication, but most of the tests are pretty concise so hopefully it isn't a big deal.

My main question to the reviewer is about what the best practice would be for using doctest. The doctest source is already included in nextSIM-DG, so we probably don't want to just copy-and-paste it here, too (as I've done). The vast majority of the "lines changed" in this PR are due to adding in the doctest source.

### CODE REVIEW

**Comment by jwallwork23** [comment]
@TomMelt In [8f33fa4](https://github.com/nextsimhub/domain_decomp/pull/62/commits/8f33fa4f0c8e1e9dd60a700a732bcb2cbfe14a18) I simply updated the pinned netcdf-c version to 4.9.2 and this worked fine. In later commits I tried removing the version pins entirely, but got errors of the form
```
Run cmake --build build --config Release
[  2%] Building CXX object CMakeFiles/domain_decomp.dir/Grid.cpp.o
[  5%] Building CXX object CMakeFiles/domain_decomp.dir/DomainUtils.cpp.o
[  8%] Building CXX object CMakeFiles/domain_decomp.dir/Partitioner.cpp.o
[ 11%] Building CXX object CMakeFiles/domain_decomp.dir/ZoltanPartitioner.cpp.o
[ 14%] Linking CXX shared library libdomain_decomp.so
[ 14%] Built target domain_decomp
[ 17%] Building CXX object CMakeFiles/decomp.dir/DomainUtils.cpp.o
[ 20%] Building CXX object CMakeFiles/decomp.dir/Grid.cpp.o
[ 23%] Building CXX object CMakeFiles/decomp.dir/Partitioner.cpp.o
[ 26%] Building CXX object CMakeFiles/decomp.dir/ZoltanPartitioner.cpp.o
[ 29%] Building CXX object CMakeFiles/decomp.dir/main.cpp.o
[ 32%] Linking CXX executable decomp
[ 32%] Built target decomp
[ 35%] Building CXX object test/CMakeFiles/test_grid_0.dir/test_grid_0.cpp.o
[ 38%] Building CXX object test/CMakeFiles/test_grid_0.dir/main.cpp.o
[ 41%] Linking CXX executable test_grid_0
[ 41%] Built target test_grid_0
[ 44%] Building CXX object test/CMakeFiles/test_grid_1.dir/test_grid_1.cpp.o
[ 47%] Building CXX object test/CMakeFiles/test_grid_1.dir/main.cpp.o
[ 50%] Linking CXX executable test_grid_1
[ 50%] Built target test_grid_1
[ 52%] Building CXX object test/CMakeFiles/test_grid_2.dir/test_grid_2.cpp.o
[ 55%] Building CXX object test/CMakeFiles/test_grid_2.dir/main.cpp.o
[ 58%] Linking CXX executable test_grid_2
[ 58%] Built target test_grid_2
[ 61%] Building CXX object test/CMakeFiles/test_zoltan_0.dir/test_zoltan_partitioner_0.cpp.o
[ 64%] Building CXX object test/CMakeFiles/test_zoltan_0.dir/main.cpp.o
[ 67%] Linking CXX executable test_zoltan_0
[ 67%] Built target test_zoltan_0
[ 70%] Building CXX object test/CMakeFiles/test_zoltan_1.dir/test_zoltan_partitioner_1.cpp.o
[ 73%] Building CXX object test/CMakeFiles/test_zoltan_1.dir/main.cpp.o
[ 76%] Linking CXX executable test_zoltan_1
[ 76%] Built target test_zoltan_1
[ 79%] Building CXX object test/CMakeFiles/test_zoltan_2.dir/test_zoltan_partitioner_2.cpp.o
[ 82%] Building CXX object test/CMakeFiles/test_zoltan_2.dir/main.cpp.o
[ 85%] Linking CXX executable test_zoltan_2
[ 85%] Built target test_zoltan_2
[ 88%] Generating test_0.nc
ncgen: error while loading shared libraries: libpnetcdf.so.4: cannot open shared object file: No such file or directory
```
when building `domain_decomp`.

**Comment by jwallwork23** [comment]
I tried building the nextSIM-DG devenv Docker image (which builds the latest version of `domain_decomp`) locally. I had to take out XIOS because of https://github.com/nextsimhub/nextsimdg/issues/719. However, `domain_decomp` built fine with the Spack environment there.

So I think once the XIOS download issue is resolved, we should be able to use the nextSIM-DG Docker image for CI in this repo, too.

File: .github/workflows/actions.yml [hunk 1]
```diff
@@ -19,63 +19,31 @@ jobs:
...
-          cmake \
-          -DENABLE_PNETCDF=ON \
-          -DCMAKE_C_COMPILER=mpicc \
-          -DCMAKE_CXX_COMPILER=mpicxx \
-          -DCMAKE_BUILD_TYPE=Release \
-          -DENABLE_TESTS=OFF \
-          -DENABLE_PARALLEL_TESTS=OFF \
-          -Bbuild -S.
-          sudo cmake --build build --config Release --target install
-          cd ..
```
Comment (TomMelt) [inline .github/workflows/actions.yml:65]: It's so nice to remove all of this :sweat_smile:

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -13,28 +13,30 @@ function(create_test)
...
-    PRIVATE Catch2::Catch2 ${PROJECT_NAME}::${LIB_NAME}
+    PRIVATE doctest::doctest ${PROJECT_NAME}::${LIB_NAME}
     )
   target_include_directories(${ARG_TARGET}
-    PRIVATE ${CMAKE_SOURCE_DIR}
+    PRIVATE ${CMAKE_SOURCE_DIR} ${DOCTEST_INCLUDE_DIR}
     )
-  foreach(numProcs RANGE 1 ${ARG_MAX_NUM_PROCS})
-    add_test(NAME ${ARG_TARGET}_n${numProcs} COMMAND ${MPIEXEC} --oversubscribe ${MPIEXEC_NUMPROC_FLAG} ${numProcs} ${MPIEXEC_PREFLAGS} ./${ARG_TARGET})
-  endforeach()
```
Comment (TomMelt) [inline test/CMakeLists.txt:28]: what's the reason for removing this part?

I have just realised we could use something like this in nextsimdg (if it works with `doctest`)

File: test/test_zoltan_partitioner_0_MPI2.cpp [hunk 3]
```diff
@@ -25,30 +25,19 @@ TEST_CASE("ZoltanPartitioner: all land", "[zoltan]")
     // Partition grid
     partitioner->partition(*grid);
 
-    int mpi_size;
-    MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
-    assert(mpi_size <= 2);
-
     int mpi_rank;
     MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
```
Comment (TomMelt) [inline test/test_zoltan_partitioner_0_MPI2.cpp:29]: I appreciate you didn't add this in, but we could use `doctest`s built-in `test_rank` to get the rank and remove the need for using `MPI_Comm_rank` (see [here](https://github.com/doctest/doctest/blob/master/doc/markdown/extensions.md#mpi_test_case) for more info)

**Review by TomMelt — CHANGES_REQUESTED** [review]
Thanks Joe. It works but I think we can simplify it a bit to recombine the test cases. Happy to discuss in more detail.

File: .github/workflows/actions.yml (hunk 1 above)
Comment (jwallwork23) [reply .github/workflows/actions.yml:65]: Agreed!

File: test/CMakeLists.txt (hunk 2 above)
Comment (jwallwork23) [reply test/CMakeLists.txt:28]: Ah I'd been removing `MAX_NUM_PROCS` stuff but this part would potentially be useful. Will try adding back in.

**Review by TomMelt — APPROVED** [review]

**Pull request merged**